File: Trigger/TriggerCommon/TriggerMenu/python/l1/CaloInfo.py
# Copyright (C) 2002-2017 CERN for the benefit of the ATLAS collaboration

from Lvl1MenuUtil import log
from Lvl1MenuUtil import oldStyle
from Limits import CaloLimits

class Isolation:

    class Parametrization:
        def __init__(self, isobit, offset, slope, mincut, upperlimit, etamin, etamax, priority):
            self.isobit     = int(isobit)
            self.offset     = int(offset)
            self.slope      = int(slope)
            self.upperlimit = int(upperlimit)
            self.etamin     = int(etamin)
            self.etamax     = int(etamax)
            self.priority   = int(priority)
            self.mincut     = int(mincut)  # minimum value for the isolation cut
            if self.isobit not in range(1,6):
                raise RuntimeError("Isolation bit is '%i' must be between 1 and 5" % self.isobit)

        def xml(self, ind=1, step=2):
            s = ind * step * ' ' + '<Parametrization isobit="%i" offset="%i" slope="%i" mincut="%i" upperlimit="%i" etamin="%i" etamax="%i" priority="%i"/>\n' % \
                ( self.isobit, self.offset, self.slope, self.mincut, self.upperlimit, self.etamin, self.etamax, self.priority )
            return s

    class IsoGroup:
        def __init__(self):
            self.parameterizations = []

        def __iter__(self):
            for x in self.parameterizations:
                yield x

        def addIsolation(self, isobit, offset, slope, mincut=0, upperlimit=CaloLimits.ClusterOff, etamin=-49, etamax=49, priority=0):
            self.parameterizations += [ Isolation.Parametrization(isobit=isobit, offset=offset, slope=slope, mincut=mincut, upperlimit=upperlimit, etamin=etamin, etamax=etamax, priority=priority) ]
            return self


    def __init__(self):
        self.isolation = {'EMIsoForEMthr': Isolation.IsoGroup(), 'HAIsoForEMthr': Isolation.IsoGroup(), 'EMIsoForTAUthr': Isolation.IsoGroup()}

    def __getitem__(self,k):
        if not k in self.isolation:
            raise RuntimeError("Isolation group %s doesn't exist. Should be one of %s" % (k, ','.join(self.isolation.keys())) )
        return self.isolation[k]
        
    def xml(self, ind=1, step=2):
        s=""
        for thrtype in self.isolation.keys():
            s += ind * step * ' ' + '<Isolation thrtype="%s">\n' % thrtype
            for p in self.isolation[thrtype]:
                s += p.xml(ind+1,step)
            s += ind * step * ' ' + '</Isolation>\n'
        return s



class MinimumTOBPt(object):
    def __init__(self, thrtype, ptmin, window=8, etamin=-49, etamax=49, priority=0):
        self.thrtype    = str(thrtype).upper()
        self.ptmin      = int(ptmin)
        self.etamin     = int(etamin)
        self.etamax     = int(etamax)
        self.priority   = int(priority)
        self.window     = int(window)
        if not self.thrtype in [ "JETS", "JETL", "EM", "TAU"]:
            raise RuntimeError("Unknown type %s for Minimum TOB PT, must be JETS, JETL, EM, or TAU")
        

    def xml(self, ind=1, step=2):
        window = ' window="%i"' % self.window if self.thrtype.startswith("JET") else ""
        s = ind * step * ' ' + '<MinimumTOBPt thrtype="%s"%s ptmin="%i" etamin="%i" etamax="%i" priority="%i"/>\n' % \
             ( self.thrtype, window, self.ptmin, self.etamin, self.etamax, self.priority )
        return s

class CaloInfo:
            
    AverageOfThr = 1
    LowerThr = 2
    HigherThr = 3
    def __init__(self, name='standard', globalEmScale=2, globalJetScale=1, jetWeights=[0]*12):
        self.name        = name
        self.globalEmScale  = globalEmScale
        self.globalJetScale = globalJetScale
        self.jetWeights  = jetWeights
        self.xsParams    = { 'xsSigmaScale' : 0,
                             'xsSigmaOffset': 0,
                             'xeMin'        : 0,
                             'xeMax'        : 0,
                             'teSqrtMin'    : 0,
                             'teSqrtMax'    : 0 }
        self.isolation = Isolation()
        self.minTOBPt = []

    def setName(self, name):
        self.name = name

    def setGlobalEmScale(self, x):
        self.globalEmScale = x

    def setGlobalJetScale(self, x):
        self.globalJetScale = x

    def setJetWeights(self, weights):
        self.jetWeights = weights

    def setXsParams( self, **args ):
        for k in args:
            if not k in self.xsParams:
                raise RuntimeError, "'%s' is not a MET significance parameter" %  k
            self.xsParams[k]  = args[k]

    def xml(self, ind=1, step=2):
        s = ind * step * ' ' + '<CaloInfo name="CaloInfo" global_em_scale="%g" global_jet_scale="%g">\n' % (self.globalEmScale, self.globalJetScale)
        if oldStyle():
            for i, jw in enumerate(self.jetWeights):
                s += (ind+1) * step * ' ' + '<JetWeight num="%i" weight="%i">%i</JetWeight>\n' % (i+1, jw, jw)
        s += (ind+1) * step * ' ' + '<METSignificance xeMin="11" xeMax="63" teSqrtMin="4" teSqrtMax="63" xsSigmaScale="1150" xsSigmaOffset="1640"/>\n'
        if not oldStyle():
            s += self.isolation.xml(ind+1)
            for minObjPt in self.minTOBPt:
                s += minObjPt.xml(ind+1)

        s += ind * step * ' ' + '</CaloInfo>\n'
        return s



    @staticmethod
    def calcJetWeightsFromThresholds(triggerConfigLvl1, use_fj=False):

        def thr2weights(thrs, factor=1.25):
            # thrs = [10.0, 15.0, 1023.0, 20.0, 30.0, 50.0, 75.0, 350.0]

            # allow for thresholds being out of order
            # need to skip any turned-off thresholds (thr=1023)
            sthrs = sorted( filter(lambda x: x<1023, thrs)) + [0]
            
            # k = [(10.0, 15.0), (15.0, 20.0), (20.0, 30.0), (30.0, 50.0), (50.0, 75.0), (75.0, 350.0), (350.0, 0)]
            k = zip(sthrs,sthrs[1:])

            # compute weights
            # TODO: this formula description, copied from LVL1.py doesn't seem to match the implementation (need to check with L1Calo - Joerg
            # -------------------------------------------------------------------
            # Jet thresholds: [5, 10, 18, 23, 35, 42, 70, 100 ][10, 18, 26, 65 ]
            # Jet weights(w):  [ 7, 14, 20, 29, 38, 56, 85, 100 ][9, 6, 23, 19]
            # Jet weights(W):[ 5,  6, 6, 8, 9, 17, 29, 15 ][9, 6, 23, 19]
            # W_i=0.5*(thr_{i+1}-thr_{i-1}),
            #     thr_{0}=0 and thr_{n+1}=thr_{n}+(thr_{n}-thr_{n-1})
            # --------------------------------------------------------------------
            sum = 0
            weights = dict([(t,0) for t in thrs])
            for thr,thrNext in k:
                if thrNext != 0:
                    w = int( (thr + thrNext + 1)/2.0) - sum
                else:  
                    w = int( thr * factor - sum)
                weights[ thr ] = w
                sum += w

            return [ weights[thr] for thr in thrs ]

    
        l1_thrs = triggerConfigLvl1.menu.thresholds
        jet_thresholds  = l1_thrs.allThresholdsOf('JET')
        fjet_thresholds = l1_thrs.allThresholdsOf('JF')
        log.debug('N jet thresholds: %d (8 expected)' % len(jet_thresholds))
        log.debug('N fjet thresholds: %d (4 expected)' % len(fjet_thresholds))

        jthr, fjthr           = [1023]*8, [1023]*4
        jet_names, fjet_names = ['---']*8, ['---']*4

        for j in jet_thresholds:
            log.debug('jet threshold %s: %d' % (j.name, j.thresholdInGeV()))
            jthr[j.mapping] = j.thresholdInGeV()
            jet_names[j.mapping] = j.name

        for j in fjet_thresholds:
            tvalues = j.thresholdValues
            priority = 0
            threshold_value = 0
            for tv in tvalues:
                if tv.priority > priority:
                    threshold_value = tv.value
            log.debug('fjet threshold %s: %d' % (j.name, threshold_value))
            fjthr.append(threshold_value)
            fjthr[j.mapping] = threshold_value
            fjet_names[j.mapping] = j.name
        #

        log.debug('Jet thresholds: %s' % ' '.join(jet_names))
        log.debug('Fjet thresholds: %s' % ' '.join(fjet_names))

        if len(jthr) <= 8:
            w = thr2weights(jthr)
            jw = w
        else:
            jw = [0]*8

        w = 4 * [0]

        if use_fj and len(fjthr) <= 4:
            w = thr2weights(fjthr)
        jw.extend(w)
        return jw

File: Trigger/TrigConfiguration/TrigConfL1Data/Root/CaloInfo.cxx
/*
  Copyright (C) 2002-2018 CERN for the benefit of the ATLAS collaboration
*/

#include "TrigConfL1Data/CaloInfo.h"
#include <iostream>

using namespace std;
using namespace TrigConf;

CaloInfo::CaloInfo()
   : L1DataBaseclass(),
     m_GlobalScale( 1 ),
     m_JetWeights(),
     m_CaloSinCos(),
     m_METSigParam(),
     m_IsolationHAIsoForEMthr( 5 ),
     m_IsolationEMIsoForEMthr( 5 ),
     m_IsolationEMIsoForTAUthr( 5 ),
     m_tobEM(),
     m_tobTau(),
     m_tobJetSmall(),
     m_tobJetLarge(),
     m_JetWindowSizeSmall( 0 ),
     m_JetWindowSizeLarge( 0 )
{}
  
void
CaloInfo::addJetWeight(int jw) {
   m_JetWeights.push_back(jw);
}
 

void
CaloInfo::addCaloSinCos(const TrigConf::CaloSinCos& csc) {
   m_CaloSinCos.push_back(csc);
}

void
CaloInfo::clear() {
   m_GlobalScale = 1;
   m_CaloSinCos.clear();
   m_JetWeights.clear();

   m_IsolationHAIsoForEMthr.clear();
   m_IsolationEMIsoForEMthr.clear();
   m_IsolationEMIsoForTAUthr.clear();

   m_IsolationHAIsoForEMthr.resize(5);
   m_IsolationEMIsoForEMthr.resize(5);
   m_IsolationEMIsoForTAUthr.resize(5);

}


void
CaloInfo::setIsolation(const std::string & isothrtype, unsigned int bit, const IsolationParam & isopar) {
   if(bit<1 || bit>5) {
      cout << "Isolation bit is outside the range [1,5] : " << bit << " - not being used" << endl;
      return;
   }
   if(isothrtype=="HAIsoForEMthr") {
      m_IsolationHAIsoForEMthr[bit-1] = isopar;
   } else if(isothrtype=="EMIsoForEMthr") {
      m_IsolationEMIsoForEMthr[bit-1] = isopar;
   } else if(isothrtype=="EMIsoForTAUthr") {
      m_IsolationEMIsoForTAUthr[bit-1] = isopar;
   } else {
      cout << "Isolation threshold " << isothrtype << " is unknown [HAIsoForEMthr, EMIsoForEMthr, EMIsoForTAUthr] - not being used" << endl;
   }
}

  
void
CaloInfo::print(const std::string& indent, unsigned int /*detail*/) const {

   cout << indent <<  "================================================ " << endl; 
   cout << indent << "CaloInfo: " << endl; 
   printNameIdV(indent);
   cout << indent << "  Global em scale : " << globalEmScale() << endl;
   cout << indent << "  Global jet scale: " << globalJetScale() << endl;
   if(m_JetWeights.size()>0) {
      cout << indent << "  JetWeights:   ";
      for(unsigned int i=0; i<m_JetWeights.size(); ++i) {
         cout << i << ": " << m_JetWeights[i] << ", ";
         if((i+1)%6==0) cout << endl;
         if(i==5) cout << indent << "                ";
      }
   }
   m_METSigParam.print(indent);
   for(const IsolationParam & isop : m_IsolationHAIsoForEMthr) {
      isop.print(indent);
   }
   cout << "================================================ " << endl; 
}  

void
TrigConf::CaloInfo::writeXML(std::ostream & xmlfile, int indentLevel, int indentWidth) const {
   
   indent(xmlfile, indentLevel, indentWidth)
      << "<CaloInfo name=\""<< name() <<"\" global_em_scale=\""<< globalEmScale() <<"\" global_jet_scale=\""<< globalJetScale() <<"\">" << endl;

   bool hasJetWeights(false);
   for(int jw : m_JetWeights) // check if there is a jetweight that is not 0
      if(jw!=0) hasJetWeights = true;

   if(hasJetWeights) {
      unsigned int num(1);
      for(int jw : m_JetWeights)
         indent(xmlfile, indentLevel+1, indentWidth)
            << "<JetWeight num=\"" << num++ << "\" weight=\""<< jw <<"\"/>" << endl;
   }

   metSigParam().writeXML(xmlfile, indentLevel+1, indentWidth);

   // isolation parametrization
   indent(xmlfile, indentLevel+1, indentWidth)
      << "<Isolation thrtype=\"HAIsoForEMthr\">" << endl; 
   for(const IsolationParam & isop : m_IsolationHAIsoForEMthr)
      isop.writeXML(xmlfile, indentLevel+2, indentWidth);
   indent(xmlfile, indentLevel+1, indentWidth)
      << "</Isolation>" << endl; 

   indent(xmlfile, indentLevel+1, indentWidth)
      << "<Isolation thrtype=\"EMIsoForEMthr\">" << endl; 
   for(const IsolationParam & isop : m_IsolationEMIsoForEMthr)
      isop.writeXML(xmlfile, indentLevel+2, indentWidth);
   indent(xmlfile, indentLevel+1, indentWidth)
      << "</Isolation>" << endl; 

   indent(xmlfile, indentLevel+1, indentWidth)
      << "<Isolation thrtype=\"EMIsoForTAUthr\">" << endl;
   for(const IsolationParam & isop : m_IsolationEMIsoForTAUthr)
      isop.writeXML(xmlfile, indentLevel+2, indentWidth);
   indent(xmlfile, indentLevel+1, indentWidth)
      << "</Isolation>" << endl; 


   indent(xmlfile, indentLevel+1, indentWidth)
      << "<MinimumTOBPt thrtype=\"EM\" ptmin=\"" << m_tobEM.ptmin << "\" etamin=\"" << m_tobEM.etamin << "\" etamax=\"" << m_tobEM.etamax << "\" priority=\"" << m_tobEM.priority <<  "\"/>" << endl;
   indent(xmlfile, indentLevel+1, indentWidth)
      << "<MinimumTOBPt thrtype=\"TAU\" ptmin=\"" << m_tobTau.ptmin << "\" etamin=\"" << m_tobTau.etamin << "\" etamax=\"" << m_tobTau.etamax << "\" priority=\"" << m_tobTau.priority <<  "\"/>" << endl;
   indent(xmlfile, indentLevel+1, indentWidth)
      << "<MinimumTOBPt thrtype=\"JETS\" window=\"" << m_JetWindowSizeSmall << "\" ptmin=\"" << m_tobJetSmall.ptmin << "\" etamin=\"" << m_tobJetSmall.etamin << "\" etamax=\"" << m_tobJetSmall.etamax << "\" priority=\"" << m_tobJetSmall.priority <<  "\"/>" << endl;
   indent(xmlfile, indentLevel+1, indentWidth)
      << "<MinimumTOBPt thrtype=\"JETL\" window=\"" << m_JetWindowSizeLarge << "\" ptmin=\"" << m_tobJetLarge.ptmin << "\" etamin=\"" << m_tobJetLarge.etamin << "\" etamax=\"" << m_tobJetLarge.etamax << "\" priority=\"" << m_tobJetLarge.priority <<  "\"/>" << endl;


   indent(xmlfile, indentLevel, indentWidth) << "</CaloInfo>" << endl;

}

